packages/rs-sdk/src/sync.rs

```diff
@@ -168,4 +307,58 @@ mod test {
             assert_eq!(result.unwrap(), "Success");
         }
     }
+
+    #[derive(Debug)]
+    enum MockError {
+        Generic,
+    }
+    impl CanRetry for MockError {
+        fn can_retry(&self) -> bool {
+            true
+        }
+    }
+    /// Counter for tracking number of requests sent
+    static REQUEST_COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+    async fn retry_test_function(settings: RequestSettings) -> ExecutionResult<(), MockError> {
+        // each invocation executes two retries
+        let retries = (random::<u8>() % 5) as usize;
+        let retries = if settings.retries.unwrap_or_default() < retries {
+            settings.retries.unwrap_or_default()
+        } else {
+            retries
+        };
+
+        // we sent 1 initial request to `retries` retries
+        REQUEST_COUNTER.fetch_add(1 + retries, Ordering::Relaxed);
+
+        Err(ExecutionError {
+            inner: MockError::Generic,
+            retries,
+            address: Some(Uri::from_static("http://localhost").into()),
+        })
+    }
+
+    #[tokio::test]
+    async fn test_retry() {
+        for _ in 0..5 {
+            REQUEST_COUNTER.store(0, Ordering::Relaxed);
+            let expected_requests: usize = (random::<u8>() % 100 + 1) as usize;
```

Comment:
I agree with Rabbit, the test should be determinisitc. Just use multiple test cases